ILogger_best_practice/kb/scenario_description/defaults.md

```diff
@@ -0,0 +1,182 @@
+
+```csharp
+public partial class InstanceLoggingExample
+{
+    private readonly ILogger _logger;
+
+    public InstanceLoggingExample(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(
+        Level = LogLevel.Information,
+        Message = "Could not open socket to `{HostName}`")]
+    public partial void CouldNotOpenSocket(string hostName);
+}
+```
+
+Sometimes, the log level needs to be dynamic rather than statically built into the code. You can do this by omitting the log level from the attribute and instead requiring it as a parameter to the logging method.
+
+```csharp
+public static partial class Log
+{
+    [LoggerMessage(
+        Message = "Could not open socket to `{HostName}`")]
+    public static partial void CouldNotOpenSocket(
+        ILogger logger,
+        LogLevel level, /* Dynamic log level as parameter, rather than defined in attribute. */
+        string hostName);
+}
+```
+
+You can omit the logging message and `System.String.Empty` will be provided for the message. The state will contain the arguments, formatted as key-value pairs.
+
+Consider the example logging output when using the `JsonConsole` formatter.
+
+```json
+{
+  "LogLevel": "Information",
+  "Category": "\u003CProgram\u003EF...9CB42__SampleObject",
+  "Message": "Liana lives in Seattle.",
+  "State": {
+    "Message": "Liana lives in Seattle.",
+    "name": "Liana",
+    "city": "Seattle",
+    "{OriginalFormat}": "{Name} lives in {City}."
+  }
+}
+```
+
+## Log method anatomy
+
+The `Microsoft.Extensions.Logging.ILogger.Log` signature accepts the `Microsoft.Extensions.Logging.LogLevel` and optionally an `System.Exception`, as shown below.
+
+```csharp
+public interface ILogger
+{
+    void Log<TState>(
+        Microsoft.Extensions.Logging.LogLevel logLevel,
+        TState state,
+        System.Exception? exception,
+        Func<TState, System.Exception?, string> formatter);
+}
+```
+
+As a general rule, the first instance of `ILogger`, `LogLevel`, and `Exception` are treated specially in the log method signature of the source generator. Subsequent instances are treated like normal parameters to the message template:
+
+```csharp
+// This is a valid attribute usage
+[LoggerMessage(
+    Level = LogLevel.Debug, Message = "M1 {Ex3} {Ex2}")]
+public static partial void ValidLogMethod(
+    ILogger logger,
+    Exception ex,
+    Exception ex2,
+    Exception ex3);
+
+// This causes a warning
+[LoggerMessage(
+    Level = LogLevel.Debug, Message = "M1 {Ex} {Ex2}")]
+public static partial void WarningLogMethod(
+    ILogger logger,
+    Exception ex,
+    Exception ex2);
+```
+
+> [!IMPORTANT]
+> The warnings emitted provide details as to the correct usage of the `LoggerMessageAttribute`. In the preceding example, the `WarningLogMethod` will report a `DiagnosticSeverity.Warning` of `SYSLIB0025`.
```

Comment:
If intended for a human audience (as opposed to AI) specifying the exact error SYSLIB0025 seems unexpectedly detailed.